cmd/restic/cmd_serve.go

```diff
@@ -0,0 +1,268 @@
+	Long: `
+The serve command runs a web server to browse a repository.
+`,
+	DisableAutoGenTag: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runWebServer(cmd.Context(), serveOptions, globalOptions, args)
+	},
+}
+
+type ServeOptions struct {
+	Listen string
+}
+
+var serveOptions ServeOptions
+
+func init() {
+	cmdRoot.AddCommand(cmdServe)
+	cmdFlags := cmdServe.Flags()
+	cmdFlags.StringVarP(&serveOptions.Listen, "listen", "l", "localhost:3080", "set the listen host name and `address`")
+}
+
+type fileNode struct {
+	Path string
+	Node *restic.Node
+}
+
+func listNodes(ctx context.Context, repo restic.Repository, tree restic.ID, path string) ([]fileNode, error) {
+	var files []fileNode
+	err := walker.Walk(ctx, repo, tree, nil, func(_ restic.ID, nodepath string, node *restic.Node, err error) (bool, error) {
+		if err != nil || node == nil {
+			return false, err
+		}
+		if fs.HasPathPrefix(path, nodepath) {
+			files = append(files, fileNode{nodepath, node})
+		}
+		if node.Type == "dir" && !fs.HasPathPrefix(nodepath, path) {
+			return false, walker.ErrSkipNode
+		}
+		return false, nil
+	})
+	return files, err
+}
+
+func runWebServer(ctx context.Context, opts ServeOptions, gopts GlobalOptions, args []string) error {
+	if len(args) > 0 {
+		return errors.Fatal("this command does not accept additional arguments")
+	}
+
+	repo, err := OpenRepository(ctx, gopts)
+	if err != nil {
+		return err
+	}
+
+	if !gopts.NoLock {
+		var lock *restic.Lock
+		lock, ctx, err = lockRepo(ctx, repo)
+		defer unlockRepo(lock)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = repo.LoadIndex(ctx)
+	if err != nil {
+		return err
+	}
+
+	funcMap := template.FuncMap{
+		"FormatTime": func(time time.Time) string { return time.Format(TimeFormat) },
+	}
+	indexPage := template.Must(template.New("index").Funcs(funcMap).Parse(indexPageTpl))
+	treePage := template.Must(template.New("tree").Funcs(funcMap).Parse(treePageTpl))
+
+	http.HandleFunc("/tree/", func(w http.ResponseWriter, r *http.Request) {
+		snapshotID, curPath, _ := strings.Cut(r.URL.Path[6:], "/")
+		curPath = "/" + strings.Trim(curPath, "/")
+		_ = r.ParseForm()
+
+		sn, err := restic.FindSnapshot(ctx, repo.Backend(), repo, snapshotID)
+		if err != nil {
+			http.Error(w, "Snapshot not found: "+err.Error(), http.StatusNotFound)
+			return
+		}
+
+		files, err := listNodes(ctx, repo, *sn.Tree, curPath)
+		if err != nil || len(files) == 0 {
+			http.Error(w, "Path not found in snapshot", http.StatusNotFound)
+			return
+		}
+
+		if r.Form.Get("action") == "dump" {
+			var tree restic.Tree
+			for _, file := range files {
+				for _, name := range r.Form["name"] {
+					if name == file.Node.Name {
+						tree.Nodes = append(tree.Nodes, file.Node)
+					}
+				}
+			}
+			if len(tree.Nodes) > 0 {
+				filename := strings.ReplaceAll(strings.Trim(snapshotID+curPath, "/"), "/", "_") + ".tar.gz"
+				w.Header().Set("Content-Disposition", "attachment; filename="+filename)
```

Comment:
[The `filename` parameter in `Content-Disposition` requires quotes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition). In this case, I suggest to change the code in this way:
```suggestion
				w.Header().Set("Content-Disposition", "attachment; filename=\""+filename+"\"")
```